### ISSUE TITLE
Electron is spawning so many clients that it reaches the limit for my linux desktop

### ISSUE THREAD

**Issue by gmmeyer** [atom/electron#2873]
I have only a handful of apps open, including the slack app, atom, two terminals, chrome and a few other assorted things. Yet, and this is a recent things, not infrequently, I am prevented from opening any more because X will inform me that I have reached the maximum number of clients.

So, let's look at what's open on my computer and using a socket:

```
./~ ❯❯❯ lsof -U | tr -s ' ' | cut -d ' ' -f 1 | sort | uniq -c | sort -rns
    145 chrome
    117 slack
     57 atom
     47 dbus-daem
     11 ibus-daem
     11 tmux
     10 pulseaudi
      9 dropbox
      7 pidgin
      6 pasystray
      6 skype
      5 at-spi2-r
      5 ibus-ui-g
      5 light-loc
      5 systemd
      5 xfce4-pow
      5 xfce4-vol
      4 xfsetting
      3 at-spi-bu
      3 awesome
      3 dconf-ser
      3 emacs
      3 gvfsd
      3 ibus-dcon
      3 ibus-x11
      3 xfconfd
      2 dbus-laun
      2 gconfd-2
      2 ibus-engi
      2 syndaemon
      1 COMMAND
      1 gvfsd-fus
      1 nacl_help
      1 unclutter
      1 xclip
```

That Atom instance only has three files open in it, yet it's using 57 sockets. Slack is doing the same thing. I have like 20 tabs open in Chrome, so that number is a bit more reasonable, but it's still a lot.

So I have to ask myself, what's going on here? I don't have a good answer, but I should be able to have more than two Electron based apps open on my system without it preventing me from opening a file browser or another terminal.

**Comment by paulcbetts** [comment]
We're seeing this in the Slack Linux client as well

**Comment by paulcbetts** [comment]
Some of the handles in the Slack app are because of a bug I'm going to fix momentarily in system-idle-time, but there's probably still an issue in Electron

**Comment by zcbenz** [comment]
A blank window of Electron opens 36 unix socket files on my machine, while a blink window of chromium-browser opens 44 files.

I think the answer is Chromium's multiple process architecture, an instance of Electron has at least 3 processes open, if the crash-reporter is used then there is 4, and they have to communicate with each other with IPC. Also the main process opens about 30 threads for various things like doing IO and watching networking configs, some of them may have a pool of handles for caching.

Atom take a few more handles because it needs to watch your project's root for filesystem events, plus a few workers doing syntax highlighting, spell checking, etc.

So adding all of them up I think 36 is a reasonable number for Electron, and 56 is not that bad for Atom. And since Electron opens the same level of numbers of handles with Chromium, I think this is expected. You can probably go to http://crbug.com to report this to Chromium's issues tracker.

**Comment by paulcbetts** [comment]
I think the problem isn't that there are a lot of them, it's that they might leak over time and eventually you start seeing 'X will inform me that I have reached the maximum number of clients.' - this might _solely_ be a bug on the Slack app's end though.

**Comment by gmmeyer** [comment]
@paulcbetts is right. when I hit the maximum number, it stops me from creating any new windows at all--for anything. I get this repeated over and over again in my .xsession-errors until I close out a few things:

```
Maximum number of clients reached
Can't open display.
```

I can see why this could be seen as a reasonable number. The question of whose bug it is may be reasonable, @zcbenz, this has been reported to Chromium's issue tracker a few times.

However, Chromium seems to keep the number of sockets they have open under a bit better control than either Atom or Slack--I've seen Atom get a lot higher than 57. And, when I hit that max number, Chromium doesn't have a problem closing. Electron processes don't always close gracefully when that happens, which is why I'm not sure this is a purely Chromium problem.

But, again, I don't really know the internals all that well, so I could be wrong.

**Comment by bundyo** [comment]
I've also been bitten by this bug - also in the Slack app. It seems to be leaking Xorg connections over time. I suspect the opening and closing the app through the tray icon is doing this, but it was rather inconsistent. In addition every desktop notification leaks one Xorg connection - this one is every time. 

You can monitor the Xorg connections through:
lsof -p [PID] | grep /tmp/.X11-unix/X0 | wc -l

Closing the app frees them, but not before that.

**Comment by zcbenz** [comment]
> In addition every desktop notification leaks one Xorg connection - this one is every time.

This looks like a bug of Electron, can you create a new issue for it?

**Comment by bundyo** [comment]
I think it is the same issue, but the OP couldn't explain it :) Anyway, will create a new one.

**Comment by bundyo** [comment]
Done:
https://github.com/atom/electron/issues/2922

**Comment by gmmeyer** [comment]
Thanks, @bundyo. I'm glad someone was better able to explain this than I was. :)

**Comment by danemco** [comment]
I'm experiencing this problem too.

**Comment by bundyo** [comment]
Should be fixed in 0.33.x and Slack 1.2.4.
